source/geometry_model/box.cc

```diff
@@ -316,16 +317,60 @@ namespace aspect
     bool
     Box<dim>::point_is_in_domain(const Point<dim> &point) const
     {
-      AssertThrow(!this->get_parameters().mesh_deformation_enabled ||
-                  this->simulator_is_past_initialization() == false,
-                  ExcMessage("After displacement of the free surface, this function can no longer be used to determine whether a point lies in the domain or not."));
+      // If mesh deformation occurs over time, and we're past the
+      // first timestep (during which no mesh deformation is applied),
+      // we can no longer use the original extents of the model domain
+      // to check whether the point lies in the domain.
+      // Initial mesh deformation is applied during the first
+      // timestep, but this function might be called before the
+      // simulator is initialized and thus before we can ask whether initial
+      // or time-dependent mesh deformation is applied.
+      // Therefore, when any mesh deformation is enabled,
+      // we loop over all the current grid cells to see if the point
+      // lies within the domain.
+      // Do note that this function can be called before mesh deformation
+      // is applied in the first timestep, and therefore there is no guarantee
+      // that the point will lie in the domain after initial mesh deformation.
+      // For example, boundary traction plugins are initialized before
+      // mesh deformation plugins and the initial_lithostatic_pressure
+      // plugin calls this function while the simulator is being initialized.
+      // However, if the simulator has not been initialized,
+      // no mesh deformation has been applied and we can use the geometry model's
+      // extents to check whether a point lies in the domain.
+      if (this->get_parameters().mesh_deformation_enabled &&
+          this->simulator_is_past_initialization())
+        {
+          std::pair<const typename parallel::distributed::Triangulation<dim>::active_cell_iterator,
+              Point<dim>> it =
+                GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), point);
+
+          if (it.first.state() == IteratorState::valid)
+            return true;
+          else
+            return false;
+        }
+
+      // The maximal extents of the unperturbed box domain.
```

Comment:
Could you put the rest of this function into an `else` block to make it easier to see that this code is only executed if the `if` above was not entered?